include/aws/common/byte_buf.h

```diff
@@ -35,5 +36,49 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t c_str[],
     return (struct aws_byte_buf){.buffer = (uint8_t *)&c_str[0], .len = len};
 }
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * No copies, no string allocations. Fills in output with a list of aws_byte_buf instances where buffer is
+ * an offset into the input_str and len is the length of that string in the original buffer.
+ *
+ * Edge case rules are as follows:
+ * if the string begins with split_on, an empty string will be the first entry in output
+ * if the input string has two adjacent split_on tokens, an empty string will be inserted into the output.
+ * if the input string ends with split_on, it will be ignored.
+ *
+ * It is the user's responsibility to properly initialize output. Recommended number of preallocated elements from output
+ * is your most likely guess for the upper bound of the number of elements resulting from the split.
+ *
+ * It is the user's responsibility to make sure the input buffer stays in memory long enough to use the results.
+ */
+AWS_COMMON_API int aws_string_split_on_char(struct aws_byte_buf *input_str, char split_on,
+                                            struct aws_array_list *output);
+
+/**
+ * No copies, no string allocations. Fills in output with a list of aws_byte_buf instances where buffer is
+ * an offset into the input_str and len is the length of that string in the original buffer.
+ *
+ * Edge case rules are as follows:
+ * if the string begins with split_on, an empty string will be the first entry in output
+ * if the input string has two adjacent split_on tokens, an empty string will be inserted into the output.
+ * if the input string ends with split_on, it will be ignored.
+ *
+ * It is the user's responsibility to properly initialize output. Recommended number of preallocated elements from output
+ * is your most likely guess for the upper bound of the number of elements resulting from the split.
+ *
+ * It is the user's responsibility to make sure the input buffer stays in memory long enough to use the results.
+ *
+ * We assume split_on is a literal and will call strlen() on it.
```

Comment:
What about binary strings?

Reply:
it wasn't really designed for that use case, but I don't see any reason we couldn't change that interface to use aws_byte_buf for split_on instead of a const char *